File: src/physics.rs
use sdl2::mixer::Chunk;
use crate::player::Player;
use crate::tile::Tile;

pub fn handle_collisions(player: &mut Player, tiles: &mut Vec<Tile>, move_fx: &Chunk) {
    handle_collision_x(player, tiles, move_fx);
    handle_collision_y(player, tiles, move_fx);
}

pub fn does_intersect(player: &mut Player, tile: &mut Tile) -> bool {
    (player.x < tile.x + tile.width as i32)
        && (player.x + player.width as i32 > tile.x)
        && (player.y + 25 < tile.y + tile.height as i32)
        && (player.y + player.height as i32 > tile.y)
}

fn handle_collision_x(player: &mut Player, tiles: &mut Vec<Tile>, move_fx: &Chunk) {
    if player.velx != 0 {
        player.move_player(player.velx, 0);
        for tile in tiles {
            if tile.isblock || tile.iswall {
                if does_intersect(player, tile) {
                    let dir: i32 = if player.velx > 0 { -1 } else { 1 };
                    if tile.isblock {
                        if tile.resistancex <= 0 {
                            tile.targetx = dir * -1 * tile.width as i32 + tile.x;
                            tile.resistancex = 30;
                            tile.resistancey = 30;
                        } else {
                            tile.resistancex -= 1;
                        }
                    }
                    player.x = if dir == 1 {
                        tile.x + tile.width as i32
                    } else {
                        tile.x - player.width as i32
                    };
                } else {
                    tile.resistancex = 30;
                }
            }
        }
        player.velx = 0;
    }
}

fn handle_collision_y(player: &mut Player, tiles: &mut Vec<Tile>, move_fx: &Chunk) {
    if player.vely != 0 {
        player.move_player(0, player.vely);
        for tile in tiles {
            if tile.isblock || tile.iswall {
                if does_intersect(player, tile) {
                    let dir: i32 = if player.vely > 0 { -1 } else { 1 };
                    if tile.isblock {
                        if tile.resistancey <= 0 {
                            tile.targety = dir * -1 * tile.height as i32 + tile.y;
                            tile.resistancex = 30;
                            tile.resistancey = 30;
                        } else {
                            tile.resistancey -= 1;
                        }
                    }
                    player.y = if dir == 1 {
                        tile.y - 25  + tile.height as i32
                    } else {
                        tile.y - player.height as i32
                    };
                } else {
                    tile.resistancey = 30;
                }
            }
        }
        player.vely = 0;
    }
}

File: src/physics.go
package mindseye

func DoesIntersect(player *Player, tile *Tile) bool {
	return player.pos.X < tile.bb.X+tile.bb.Width &&
		player.pos.X+float32(player.width) > tile.bb.X &&
		player.pos.Y+25 < tile.bb.Y+tile.bb.Height &&
		player.pos.Y+float32(player.height) > tile.bb.Y
}

func HandleCollisions(player *Player, tiles []Tile, dt float32) {
	handleCollisionX(player, tiles, dt)
	handleCollisionY(player, tiles, dt)
}

func handleCollisionX(player *Player, tiles []Tile, dt float32) {
	if player.vel.X == 0 {
		return
	}

	player.pos.X += player.vel.X
	for i := range tiles {
		tile := &tiles[i]
		if !tile.isBlock && !tile.isWall {
			continue
		}
		if DoesIntersect(player, tile) {
			dir := float32(1)
			if player.vel.X > 0 {
				dir = -1
			}
			if tile.isBlock && !tile.isMoving {
				if tile.resistance <= 0 {
					tile.targetPos.X = tile.bb.X - dir*tile.bb.Width
					tile.resistance = 30
				} else {
					tile.resistance -= dt * 200
				}
			}
			if dir == 1 {
				player.pos.X = tile.bb.X + tile.bb.Width
			} else {
				player.pos.X = tile.bb.X - float32(player.width)
			}
		} else {
			tile.resistance = 30
		}
	}
	player.vel.X = 0
}

func handleCollisionY(player *Player, tiles []Tile, dt float32) {
	if player.vel.Y == 0 {
		return
	}

	player.pos.Y += player.vel.Y
	for i := range tiles {
		tile := &tiles[i]
		if !tile.isBlock && !tile.isWall {
			continue
		}
		if DoesIntersect(player, tile) {
			dir := float32(1)
			if player.vel.Y > 0 {
				dir = -1
			}
			if tile.isBlock && !tile.isMoving {
				if tile.resistance <= 0 {
					tile.targetPos.Y = tile.bb.Y - dir*tile.bb.Height
					tile.resistance = 30
				} else {
					tile.resistance -= dt * 200
				}
			}
			if dir == 1 {
				player.pos.Y = -25 + tile.bb.Y + tile.bb.Height
			} else {
				player.pos.Y = tile.bb.Y - float32(player.height)
			}
		} else {
			tile.resistance = 30
		}
	}
	player.vel.Y = 0
}
